openedx/core/djangoapps/call_stack_manager/core.py

```diff
@@ -1,144 +1,203 @@
+
     Returns:
         wrapped function
     """
+    @wrapt.decorator
+    def real_donottrack(wrapped, instance, args, kwargs):  # pylint: disable=W0613
+        """ Takes function to be decorated and returns wrapped function
 
-    def real_donottrack(function):
-        """takes function to be decorated and returns wrapped function
+        Arguments:
+            wrapped - The wrapped function which in turns needs to be called by wrapper function.
+            instance - The object to which the wrapped function was bound when it was called.
+            args - The list of positional arguments supplied when the decorated function was called.
+            kwargs - The dictionary of keyword arguments supplied when the decorated function was called.
 
-        Args:
-            function - wrapped function i.e. real_donottrack
         """
-        def wrapper(*args, **kwargs):
-            """ wrapper function for decorated function
-            Returns:
-                wrapper function i.e. wrapper
-            """
-            if len(classes_not_to_be_tracked) == 0:
-                global TRACK_FLAG  # pylint: disable=W0603
-                current_flag = TRACK_FLAG
-                TRACK_FLAG = False
-                function(*args, **kwargs)
-                TRACK_FLAG = current_flag
+        if entities_not_to_be_tracked:
+            global HALT_TRACKING  # pylint: disable=W0603
+            HALT_TRACKING.append(entities_not_to_be_tracked)
+            HALT_TRACKING[-1] = list(set([x for sublist in HALT_TRACKING for x in sublist]))
+            return_value = wrapped(*args, **kwargs)
+
+            # check if the returning class is a generator
+            if isinstance(return_value, types.GeneratorType):
+                def generator_wrapper(wrapped_generator):
+                    try:
+                        while True:
+                            return_value = next(wrapped_generator)
+                            yield return_value
+                    finally:
+                        global HALT_TRACKING
+                        HALT_TRACKING.pop()
+                return generator_wrapper(return_value)
+            else:
+                HALT_TRACKING.pop()
+                return return_value
+
+        else:  # if donottrack is not parameterized
+            global TRACK_FLAG  # pylint: disable=W0603
+            TRACK_FLAG = False
+            return_value = wrapped(*args, **kwargs)
+
+            # check if the returning class is a generator
+            if isinstance(return_value, types.GeneratorType):
+                def generator_wrapper(wrapped_generator):
+                    try:
+                        while True:
+                            return_value = next(wrapped_generator)
+                            yield return_value
+                    finally:
+                        global TRACK_FLAG
+                        TRACK_FLAG = True
+                return generator_wrapper(return_value)
```

Comment:
You could avoid all of this duplication by just having a special object (perhaps `None`, or perhaps a constant `object()`) that you put on the `HALT_TRACKING` stack that means "don't track anything". Then you'd just have to check the top of the stack, rather than having to check two different variables anywhere you need to verify that you shouldn't be tracking.